### ISSUE TITLE
Inconsistent Transferable Balance Calculation Logic

### ISSUE THREAD

**Issue by ahmadkaouk** [paritytech/polkadot-sdk#1833]
An inconsistency has been identified in the Polkadot SDK concerning the logic used for computing transferable balances. This divergence may mislead users about the actual spendable balance, potentially leading to transaction failures due to inadequate fee coverage.

The inconsistency arises from the utilization of different formulas:
1. Old Formula: `transferable = free - max(frozen, reserved)`.
2. New Formula: `transferable = free - (frozen - reserved)`.

This originates from the `CurrencyAdapter`, which relies on the outdated `Currency` trait as illustrated  [here](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/transaction-payment/src/payment.rs#L84). The new transferable balance is computed utilizing the `reducible_balances` function defined [here](https://github.com/paritytech/polkadot-sdk/blob/e8da320734ae44803f89dd2b35b3cfea0e1ecca1/substrate/frame/balances/src/impl_fungible.rs#L44).

The discrepancy is particularly noticeable in the `eth.getBalance` method in Frontier, where the new formula is employed. Under certain circumstances, a positive balance may be displayed, indicating available funds. However, transactions may fail due to an insufficient balance to cover the fees, given that the old formula is utilized for fee calculations.

- Is the discrepancy in the logic for calculating "transferable balance" concerning fees a deliberate design or an oversight?
- Is there a rationale behind not extending the new logic to fee calculations, and if so, what is the reasoning behind this approach?

**Comment by kianenigma** [comment]
IN any case, an item missing (and rather high priority) in https://github.com/paritytech/polkadot-sdk/issues/226 is then migrate `CurrencyAdapter` to also work with `fungible`, I added it there.

**Comment by michalisFr** [comment]
@ahmadkaouk Just to point that if the account cannot be reaped (which should be the case when there's a hold and/or freeze) or we don't want it to be reaped, the transferable balance calculation is this:
`transferable = free - max(frozen - reserved, ED)`
since now the ED is a requirement over the free balance, not the total balance.

**Comment by liamaharon** [comment]
Spendable balance indeed I think should be `transferable = free - max(frozen, reserved)`, NOT `transferable = free - (frozen - reserved)` (@bkchr can you confirm?).

I will work on this.

**Comment by liamaharon** [comment]
@ahmadkaouk I have a couple of questions: 

1. Could you please clarify why you determined this regression to have originated in the `CurrencyAdaptor`?

After initial review, it seems was originated by the change in behavior of `fn reducible_balance` introduced in https://github.com/paritytech/substrate/pull/12951.

Prior implementation:

https://github.com/paritytech/polkadot-sdk/blob/c699876ab81403f914276bb0fe173b8ad9b96603/substrate/frame/balances/src/lib.rs#L1070-L1083

New implementation: 

https://github.com/paritytech/polkadot-sdk/blob/495d24d730bb9c2205a5305f383627eabda4a323/substrate/frame/balances/src/impl_fungible.rs#L44-L70

2. Where did you get this formula?

> Old Formula: transferable = free - max(frozen, reserved).

It seems only reserved and free are used in the prior implementation, I don't see usage of reserved anywhere.

Thanks.

**Comment by gavofyork** [comment]
#12951 did alter the functionality of existing reserves by design:

> Alter reserve functionality in Balances so that it does not contribute to the account's existential deposit and does use a consumer reference. Remove the concept of withdraw reasons from locks. Migration is lazy: accounts about to be mutated will be migrated prior. New accounts will automatically be mutated. A permissionless upgrade dispatchable is provided to upgrade dormant accounts.

That Frontier is now showing some inconsistencies in its `getBalance` (which presumably reports transferability rather than the real balance) indicates these repercussions have not yet been integrated. Introducing a test would probably help ensure this doesn't happen again.

**Comment by liamaharon** [comment]
@gavofyork I have a clarifying question about the `Balances` implementation of `Inspect::reducible_balance`.

Here's the doc comment describing the behavior of the method:
https://github.com/paritytech/polkadot-sdk/blob/c4211b6533bd4dbb5a77b861c1ea35d9885b4f4f/substrate/frame/support/src/traits/tokens/fungible/regular.rs#L80-L85

Here is the `Balances` implementation:

https://github.com/paritytech/polkadot-sdk/blob/495d24d730bb9c2205a5305f383627eabda4a323/substrate/frame/balances/src/impl_fungible.rs#L44-L70

Is line 54 (`untouchable = a.frozen.saturating_sub(a.reserved);`) correct, or is it supposed to be `untouchable = a.frozen.max(a.reserved);`?

Otherwise, even when `Inspect::reducible_balance` is called with `Polite` it seems it can return a value greater than `a.reserved` and `a.frozen`, which doesn't seem right given the behavior described in the doc comment.

**Comment by gavofyork** [comment]
They're both correct.

If `Polite`, then we are unable to reduce the total balance to below the frozen balance. Since we cannot touch the balance on hold anyway and since the balance on hold contributes to the total, then we can discount what is untouchable by this.

Frozen and reserved (aka on-hold)

**Comment by liamaharon** [comment]
@ahmadkaouk could you please confirm if updating the `CurrencyAdaptor` to use `fungible` (therefore the new formula) instead `Currency` should resolve this issue with Frontier?

Is there any test cases I can use to reproduce the issue?

**Comment by gavofyork** [comment]
> Visual aid:

Yes, and you can add the ED into that, too:

```
|__total__________________________________|
|__on_hold__|_____________free____________|
|__________frozen___________|__spendable__|
|__on_hold__|__ed__|        |
            |__untouchable__|

untouchable = max(ed, frozen - on_hold)
```
(or if we don't care about ED because we have other providers or will reap and don't have any consumers, then it's just `untouchable = frozen - on_hold`)

**Comment by ahmadkaouk** [comment]
> @ahmadkaouk could you please confirm if updating the `CurrencyAdaptor` to use `fungible` (therefore the new formula) instead `Currency` should resolve this issue with Frontier?
> 
> Is there any test case I can use to reproduce this?

@liamaharon, yes I believe that updating `CurrencyAdapter` to use `fungible` can solve the problem. I have created a test on the Moonbeam network that reproduces the issue [here](https://github.com/moonbeam-foundation/moonbeam/pull/2512).

Here's a summary of the scenario that triggers the problem:

1. A new account is created with an initial balance of 30 GLMR.
2. The new account submits a proposal with a deposit of `minDepositAmount` (4 GLMR). (Total of 9 GLMR are reserved)
3. The new account delegates 20 GLMR.
4. Then the account transfers 5 GLMR to another account.
5. We verify that the account has a transferable balance of approximately 5 GLMR.
6. A second transfer of 2 GLMR  is made from the new account. 
7. The transfer fails with this error message "Invalid Transaction: Inability to pay some fees , e.g. account balance too low"
 
 Balance info for the new account at the end:
```
  data: {
    free: 15,996,484,162,237,998,502
    reserved: 9,002,400,000,000,000,000
    frozen: 20,000,000,000,000,000,000
    flags: 170,141,183,460,469,231,731,687,303,715,884,105,728
  }
  ```
Note that ED is 0 in this scenario.

**Comment by liamaharon** [comment]
@ahmadkaouk I have a PR adding a `FungibleAdapter` - https://github.com/paritytech/polkadot-sdk/pull/2292. 

Could you try cherry-pick the commit and replace usage of `CurrencyAdapter` with `FungibleAdapter` in your runtime to check if it resolves the issue?

**Comment by liamaharon** [comment]
Hey @ahmadkaouk can try the `FungibleAdapter` from https://github.com/paritytech/polkadot-sdk/pull/2292?

**Comment by ggwpez** [comment]
We kind of need a unit test here in the SDK to prevent this from happening again and checking that #2292 does solve the issue.

**Comment by kianenigma** [comment]
> Ah, thank you for clarifying.
> 
> Visual aid:
> 
> ```
> |_______total___________________|
> |_______on_hold____|____free____|
> |_______frozen___________|
>                     xxxxx
>                       ^
>          untouchable = frozen - on_hold
> ```

What helped me understand this convo and digram(s) better is to specify that `untouchable` is the portion of `free` that is `untouchable`, not `total`. In other words I would call it `free_untouchable`.

**Comment by ahmadkaouk** [comment]
Hey @liamaharon, just wanted to let you know that I tested the `FungibleAdapter` from #2292 and it does fix the issue.

**Comment by liamaharon** [comment]
> Hey @liamaharon, just wanted to let you know that I tested the `FungibleAdapter` from #2292 and it does fix the issue.

@ahmadkaouk fyi this is finally merged and will be included in the next release.
